rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+
+### Process Configs
+
+Let's rethink the current design of CLI options and config files. ([#5676](https://github.com/risingwavelabs/risingwave/issues/5676)) Given the existence of system parameters, some existing parameters may need to be moved to system parameters, like `data_directory` as mentioned before.
+
+We propose to consider CLI options as a way to override config files (i.e. in term of priority, CLI options -> config file -> defaults); as a result, the items in CLI options are a subset of the config file. The reasons for this include
+
+- All configs are still too many to be written as CLI options. (See [example.toml](https://github.com/risingwavelabs/risingwave/blob/main/src/config/example.toml)) 
+- While CLI configs are handy to set some frequently-used options like endpoint and memory.
+
+This pattern is widely adopted in many databases and even other applications. For example, TiDB considers configuration file as a superset of CLI options:
+
+> The TiDB configuration file supports more options than command-line parameters. You can download the default configuration file [config.toml.example](https://github.com/pingcap/tidb/blob/master/config/config.toml.example) and rename it to `config.toml`. This document describes only the options that are not involved in [command line options](https://docs.pingcap.com/tidb/stable/command-line-flags-for-tidb-configuration).
+>
+> Ref. [TiDB Configuration File | PingCAP Docs](https://docs.pingcap.com/tidb/stable/tidb-configuration-file)
+
+In RisingWave it will be like
+
+```toml
+# risingwave.toml
+[server]
+total_memory_bytes = 1073741824
+parallelism = 4
+[storage]
+shared_buffer_capacity_mb = 4096
+sstable_size_mb = 256
+block_size_kb = 1024
+```
+
+```bash
+# Override total_memory_bytes as 2GB
+./risingwave compute-node --total-memory-bytes=2147483648
+```
+
+### Mutability
+
+- Process configs must be **immutable**, because there is no user-interface to change it.
+- Global configs might be **immutable or mutable**, the mutable ones can be further divided into take effect immediately or need reboot.
+    - **Immutable** e.g. `data_directory` should be rejected upon SET
+    - **Mutable & need reboot** e.g. `heartbeat_interval_ms`
+    - **Mutable & immediately** e.g. `batch_chunk_size`
+
+### Style
+
+- Process configs follow the style of TOML (lower-case, underscore) and ordinary CLI arguments (lower-case, hyphen)
+- Global configs should be flatten unlike TOML. Consider add a prefix e.g. `stream_foo`, `batch_foo` if it seems ambiguous.
+- Dangerous configs should be named with keyword `developer` or `unsafe` as a warning.
+
+## Discussions
+
+1. Do we need database parameters, global (instance-wide) parameters, or both?
```

Comment:
🤔 Are there real cases or just a proposal? Adding levels will make the configuration complicated. I'd say that instance-wide and object(table, source, mv)-specific parameters should be enough.

Reply:
In PG there's only a set of variables, and can be set at different levels (session, role, database, instance).

> object(table, source, mv)-specific parameters

But this seems similar to database parameters. IMHO we didn't consider such things in the RFC? 🤔

Reply:
> But this seems similar to database parameters. IMHO we didn't consider such things in the RFC? 🤔

You're right. Let's be PG-similar.

Reply:
Not decided yet. But anyway we can add support for database parameters after the completion of global parameters.